sparse_ho/criterion.py

```diff
@@ -20,178 +20,156 @@ class HeldOutMSE():
+    def __init__(
+            self, idx_train, idx_val, X_test=None, y_test=None,):
         """
         Parameters
         ----------
-        X_val : {ndarray, sparse matrix} of shape (n_samples, n_features)
-            Validation data
-        y_val : {ndarray, sparse matrix} of shape (n_samples,)
-            Validation target
-        model: object of the class Model (e.g. Lasso or SparseLogreg)
+        idx_train: np.array
+            indices of the training set
+        idx_test: np.array
+            indices of the testing set
         X_test : {ndarray, sparse matrix} of shape (n_samples_test, n_features)
             Test data
-        convexify: bool
-            this param should be remove from here XXX
-        gamma_convex: bool
-            this param should be removed from here XXX
         y_test : {ndarray, sparse matrix} of (n_samples_test)
             Test target
         """
-        self.X_val = X_val
-        self.y_val = y_val
         self.X_test = X_test
         self.y_test = y_test
-        self.model = model
-        self.convexify = convexify
-        self.gamma_convex = gamma_convex
+        self.idx_train = idx_train
+        self.idx_val = idx_val
 
         self.mask0 = None
         self.dense0 = None
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        return 2 * (self.X_val[:, mask].T @ (
-            self.X_val[:, mask] @ dense - self.y_val)) / self.X_val.shape[0]
-
-    def value(self, mask, dense):
-        val = (
-            norm(self.y_val - self.X_val[:, mask] @ dense) ** 2 /
-            self.X_val.shape[0])
-        return val
-
-    def value_test(self, mask, dense):
-        if self.X_test is not None and self.y_test is not None:
-            self.val_test = (
-                norm(self.y_test - self.X_test[:, mask] @ dense) ** 2 /
-                self.X_test.shape[0])
-        else:
-            self.val_test = None
-
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
-        else:
-            self.rmse = None
+    def value_outer_crit(self, X, y, mask, dense):
+        """Compute the MSE on the validation set.
+        """
+        if X is None or y is None:
+            return None
+        return norm(y - X[:, mask] @ dense) ** 2 / len(y)
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
         # TODO add warm start
         mask, dense, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y, log_alpha, self.model, tol=tol, compute_jac=False)
-        self.value_test(mask, dense)
-        return self.value(mask, dense)
+            X[self.idx_train], y[self.idx_train], log_alpha, model, tol=tol, compute_jac=False)
+        self.get_mse_test(mask, dense)
```

Comment:
I got confused too. This function returns MSE in estimation, norm(beta_k - beta_star) ** 2.    It should also be handled by Monitor, not criterion, and will be removed.

Reply:
I would recommend using norm(beta_k - beta_star) ** 2 /norm(beta_star) ** 2 to get something one can compare when varying n_samples or n_features. WDYT?

Reply:
> I got confused too. This function returns MSE in estimation, norm(beta_k - beta_star) ** 2. It should also be handled by Monitor, not criterion, and will be removed.
 
> I would recommend using norm(beta_k - beta_star) ** 2 /norm(beta_star) ** 2 to get something one can compare when varying n_samples or n_features. WDYT?

Yes and yes, this will be for the next PR! :)

Reply:
> as the class is called HeldOutMSE can you use a more generic name that get_mse_test? self.eval? self.compute?

The current name of this function is ```value_outer_crit```